File: cli/src/mandala_fleet/mcp/activity.py
"""Activity middleware: surface every tool call to a sink.

When the server is hosted by the TUI, the sink posts each call into the
Claude-activity pane so the operator watches what the client is doing —
the tool name, its arguments, and whether it succeeded or errored. Every
call emits TWICE: a `start` event as the tool begins (so a long eval or
survey is visible while it runs, not only after it lands) and an
`ok`/`error` event when it settles; the two share a `seq` so a watcher
can pair them even with concurrent same-named calls in flight. The settle
event also carries `elapsed` (seconds) and a small `result` summary
(ok/refused/run_id when the tool returned them) — enough for the pane to
show durations and for the TUI to attach the exact run a tool registered.
The sink must never break a tool call, so its exceptions are swallowed.

`audit_event` is the persistent trail: settled MUTATING calls are appended
to `state_dir()/mcp/audit.jsonl` regardless of transport, so a headless
stdio session leaves the same record the TUI operator watches live.
"""

from __future__ import annotations

import itertools
import json
import time
from typing import Any, Callable

from fastmcp.server.middleware import Middleware

ActivitySink = Callable[[dict[str, Any]], None]

# Tools whose settled calls land in the audit trail: everything that can
# change fleet state (or swap what the server serves).
_AUDITED = frozenset({"deploy", "reboot", "restart_service", "reload"})


def _result_summary(result: Any) -> dict | None:
    """The few result fields a watcher acts on (ok/refused/run_id), pulled
    defensively from FastMCP's ToolResult structured content."""
    data = getattr(result, "structured_content", None)
    if isinstance(data, dict) and set(data) == {"result"} and isinstance(data["result"], dict):
        data = data["result"]  # unwrapped non-object returns
    if not isinstance(data, dict):
        return None
    summary = {k: data[k] for k in ("ok", "refused", "run_id") if k in data}
    return summary or None


def audit_event(event: dict) -> None:
    """Append a settled mutating call to the per-user audit log. Best
    effort — an unwritable state dir must never sink a tool call."""
    if event.get("status") == "start" or event.get("tool") not in _AUDITED:
        return
    from ..drift import state_dir

    path = state_dir() / "mcp" / "audit.jsonl"
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        with open(path, "a", encoding="utf-8") as fh:
            fh.write(json.dumps({"ts": time.time(), **event}) + "\n")
    except (OSError, TypeError, ValueError):
        pass


class ActivityMiddleware(Middleware):
    """Emit `{tool, args, status, detail, seq}` per tool call to a sink —
    `status: "start"` at call entry, then `"ok"`/`"error"` (plus `elapsed`
    and a `result` summary) at settle."""

    def __init__(self, sink: ActivitySink) -> None:
        self._sink = sink
        self._seq = itertools.count(1)

    async def on_call_tool(self, context, call_next):  # type: ignore[no-untyped-def]
        params = context.message
        name = getattr(params, "name", "?")
        args = dict(getattr(params, "arguments", {}) or {})
        seq = next(self._seq)
        started = time.monotonic()
        self._emit({
            "tool": name, "args": args, "status": "start",
            "detail": None, "seq": seq,
        })
        try:
            result = await call_next(context)
        except Exception as e:  # noqa: BLE001 — re-raised after emitting
            self._emit({
                "tool": name, "args": args, "status": "error",
                "detail": str(e), "seq": seq,
                "elapsed": round(time.monotonic() - started, 3),
            })
            raise
        self._emit({
            "tool": name, "args": args, "status": "ok",
            "detail": None, "seq": seq,
            "elapsed": round(time.monotonic() - started, 3),
            "result": _result_summary(result),
        })
        return result

    def _emit(self, event: dict) -> None:
        try:
            self._sink(event)
        except Exception:  # noqa: BLE001 — the feed must never sink a call
            pass

File: crates/mandala-mcp/src/activity.rs
//! Activity events + the persistent audit trail.
//!
//! rust-mcp-sdk has no middleware concept, so the Python `ActivityMiddleware`
//! becomes a dispatch wrapper (see `server.rs`): every tool call funnels
//! through one `call_tool`, which emits TWICE — a `start` event as the tool
//! begins (so a long eval or survey is visible while it runs, not only after
//! it lands) and an `ok`/`error` event when it settles; the two share a `seq`
//! so a watcher can pair them even with concurrent same-named calls in
//! flight. The settle event also carries `elapsed` (seconds) and a small
//! `result` summary (ok/refused/run_id when the tool returned them) — enough
//! for a pane to show durations and for the phase-2 TUI to attach the exact
//! run a tool registered. The sink must never break a tool call.
//!
//! [`audit_event`] is the persistent trail: settled MUTATING calls are
//! appended to `state_dir()/mcp/audit.jsonl` regardless of transport, so a
//! headless stdio session leaves the same record a TUI operator watches live.

use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use mandala_core::drift::state_dir;
use serde_json::Value;

/// A pluggable receiver for activity events (`{tool, args, status, detail,
/// seq}` plus `elapsed`/`result` on settle). The phase-2 TUI feeds its
/// activity pane from this; the stdio server default is no sink. The sink is
/// called synchronously on the dispatch path — it must be cheap and must not
/// panic.
pub type ActivitySink = Arc<dyn Fn(&Value) + Send + Sync>;

/// Tools whose settled calls land in the audit trail: everything that can
/// change fleet state (or swap what the server serves).
const AUDITED: [&str; 4] = ["deploy", "reboot", "restart_service", "reload"];

/// Unix epoch seconds as a float — the audit `ts`, parity with Python
/// `time.time()`.
fn now_epoch_f64() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}

/// Append a settled mutating call to the per-user audit log
/// (`state_dir()/mcp/audit.jsonl`, one JSON object per line, `ts` first-class).
/// Best effort — an unwritable state dir must never sink a tool call.
pub fn audit_event(event: &Value) {
    let status = event.get("status").and_then(Value::as_str);
    let tool = event.get("tool").and_then(Value::as_str).unwrap_or("");
    if status == Some("start") || !AUDITED.contains(&tool) {
        return;
    }
    let Some(obj) = event.as_object() else {
        return;
    };
    let mut line = serde_json::Map::new();
    line.insert("ts".to_string(), Value::from(now_epoch_f64()));
    for (k, v) in obj {
        line.insert(k.clone(), v.clone());
    }
    let path = state_dir().join("mcp").join("audit.jsonl");
    let write = || -> std::io::Result<()> {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        use std::io::Write;
        let mut fh = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&path)?;
        writeln!(fh, "{}", Value::Object(line))?;
        fh.flush()
    };
    let _ = write();
}

/// The few result fields a watcher acts on (ok/refused/run_id), pulled from a
/// tool's structured result. `None` when the result carries none of them.
#[must_use]
pub fn result_summary(result: &Value) -> Option<Value> {
    let obj = result.as_object()?;
    let mut summary = serde_json::Map::new();
    for key in ["ok", "refused", "run_id"] {
        if let Some(v) = obj.get(key) {
            summary.insert(key.to_string(), v.clone());
        }
    }
    if summary.is_empty() {
        None
    } else {
        Some(Value::Object(summary))
    }
}
